File: src/main/java/com/hileco/cortex/instructions/math/DIVIDE.java
package com.hileco.cortex.instructions.math;

import com.hileco.cortex.context.VirtualMachine;
import com.hileco.cortex.context.ProgramContext;
import lombok.EqualsAndHashCode;

import java.math.BigInteger;

@EqualsAndHashCode(callSuper = true)
public class DIVIDE extends MathInstruction {
    @Override
    public BigInteger innerExecute(VirtualMachine process, ProgramContext program, BigInteger left, BigInteger right) {
        return left.divide(right);
    }
}

File: src/main/java/com/hileco/cortex/instructions/math/DIVIDE.kt
package com.hileco.cortex.instructions.math

import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.constraints.expressions.Expression.Divide
import com.hileco.cortex.constraints.expressions.Expression.Value
import java.math.BigInteger

class DIVIDE : MathInstruction() {
    override fun calculate(left: BigInteger, right: BigInteger): BigInteger {
        return left.divide(right)
    }

    override fun calculate(left: Expression, right: Expression): Expression {
        if (left is Value && right is Value) {
            val result = calculate(left.constant.toBigInteger(), right.constant.toBigInteger())
            return Value(result.toLong())
        }
        return Divide(left, right)
    }
}
